libs/core/langchain_core/runnables/context.py

```diff
@@ -0,0 +1,246 @@
+            return await configurable[self.ids[0]]()
+
+
+SetValue = Union[
+    Runnable[Input, Output],
+    Callable[[Input], Output],
+    Callable[[Input], Awaitable[Output]],
+    Any,
+]
+
+
+def _coerce_set_value(value: SetValue) -> Runnable[Input, Output]:
+    if not isinstance(value, Runnable) and not callable(value):
+        return coerce_to_runnable(lambda _: value)
+    return coerce_to_runnable(value)
+
+
+class ContextSet(RunnableSerializable):
+    keys: Mapping[str, Optional[Runnable]]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        key: Optional[str] = None,
+        value: Optional[SetValue] = None,
+        **kwargs: SetValue,
+    ):
+        if key is not None:
+            kwargs[key] = value
+        super().__init__(
+            keys={
+                k: _coerce_set_value(v) if v is not None else None
+                for k, v in kwargs.items()
+            }
+        )
+
+    @property
+    def ids(self) -> List[str]:
+        return [
+            f"{CONTEXT_CONFIG_PREFIX}{key}{CONTEXT_CONFIG_SUFFIX_SET}"
+            for key in self.keys
+        ]
+
+    @property
+    def config_specs(self) -> List[ConfigurableFieldSpec]:
+        mapper_config_specs = [
+            s
+            for mapper in self.keys.values()
+            if mapper is not None
+            for s in mapper.config_specs
+        ]
+        for spec in mapper_config_specs:
+            if spec.id.endswith(CONTEXT_CONFIG_SUFFIX_GET):
+                getter_key = spec.id.split("/")[1]
+                if getter_key in self.keys:
+                    raise ValueError(
+                        f"Circular reference in context setter for key {getter_key}"
+                    )
+        return super().config_specs + [
+            ConfigurableFieldSpec(
+                id=id_,
+                annotation=Callable[[], Any],
+            )
+            for id_ in self.ids
+        ]
+
+    def invoke(self, input: Any, config: Optional[RunnableConfig] = None) -> Any:
+        config = config or {}
+        configurable = config.get("configurable", {})
+        for id_, mapper in zip(self.ids, self.keys.values()):
+            if mapper is not None:
+                configurable[id_](mapper.invoke(input, config))
+            else:
+                configurable[id_](input)
+        return input
+
+    async def ainvoke(
+        self, input: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
+    ) -> Any:
+        config = config or {}
+        configurable = config.get("configurable", {})
+        for id_, mapper in zip(self.ids, self.keys.values()):
+            if mapper is not None:
+                await configurable[id_](await mapper.ainvoke(input, config))
+            else:
+                await configurable[id_](input)
+        return input
+
+
+class Context:
+    @staticmethod
+    def getter(key: Union[str, List[str]], /) -> ContextGet:
+        return ContextGet(key)
+
```

Comment:
Can we somehow make the `Context` key-value storage to be local scoped rather than globally accessible?

The reason is that, if some developer create a custom chain (either a built-in langchain runnable), or a chain from other packages published by user, etc), and a user use this custom chain, chained with their own chain with certain kinds of context involved. There might be chances of context collision, which causes code to run in a unpredictable/incorrect way. Here's an example:

```python
def test_out_of_scoped_context_variable() -> None:
    builtin_chain = seq_naive_rag()
    user_chain = {
        "leak_prev_context": Context.getter("context"),
        "set_context": Context.setter("context", RunnableLambda(lambda x: "override context")),
    } | RunnableLambda(lambda x: x["leak_prev_context"])

    rag_result = builtin_chain.invoke("What up")
    user_result = user_chain.invoke(rag_result)
    assert user_result == "override context"

    composite_chain = builtin_chain | user_chain
    composite_result = composite_chain.invoke("What up")
    assert composite_result == "override context"
```

- The actual `user_result` is expected to be "override context" since we directly invoke `user_chain` with no previous context variable injections.
- However, when composing both chain with `composite_chain`. `composite_result` becomes the `context` value passed inside the rag chain (i.e. `['Hi there!', 'How are you?', "What's your name?"]`), rather than `"override context"`, which raise an error.

Idealy, calling `builtin_chain | user_chain` as a whole or invoking each chain separately should get the same result. But because of the upscoped context injection, user defined chain would produce incorrect (or even confusing result without investigation on the definition of `rag_chain`).

I'm thinking if we can make context to be locally scoped, by dynamically altering the context prefix (maybe automatically generated by a uuid if not defined by user).